mountpoint-s3/tests/fs.rs

```diff
@@ -164,6 +166,225 @@ async fn test_read_dir_nested(prefix: &str) {
     fs.releasedir(dir_ino, dir_handle, 0).await.unwrap();
 }
 
+#[tokio::test]
+async fn test_lookup_negative_cached() {
+    let fs_config = S3FilesystemConfig {
+        cache_config: CacheConfig {
+            prefer_s3: false,
+            dir_ttl: Duration::from_secs(600),
+            file_ttl: Duration::from_secs(600),
+        },
+        ..Default::default()
+    };
+    let (client, fs) = make_test_filesystem("test_lookup_negative_cached", &Default::default(), fs_config);
+
+    let head_counter = client.new_counter(Operation::HeadObject);
+    let list_counter = client.new_counter(Operation::ListObjectsV2);
+
+    let _ = fs
+        .lookup(FUSE_ROOT_INODE, "file1.txt".as_ref())
+        .await
+        .expect_err("should fail as no object exists");
+    assert_eq!(head_counter.count(), 1);
+    assert_eq!(list_counter.count(), 1);
+
+    // Check no negative caching
+    let _ = fs
+        .lookup(FUSE_ROOT_INODE, "file1.txt".as_ref())
+        .await
+        .expect_err("should fail as no object exists");
+    assert_eq!(head_counter.count(), 2);
+    assert_eq!(list_counter.count(), 2);
+
+    client.add_object("file1.txt", MockObject::constant(0xa1, 15, ETag::for_tests()));
+
+    let _ = fs
+        .lookup(FUSE_ROOT_INODE, "file1.txt".as_ref())
+        .await
+        .expect("should succeed as object exists and no negative cache");
+    assert_eq!(head_counter.count(), 3);
+    assert_eq!(list_counter.count(), 3);
+
+    let _ = fs
+        .lookup(FUSE_ROOT_INODE, "file1.txt".as_ref())
+        .await
+        .expect("should succeed as object is cached and exists");
+    assert_eq!(head_counter.count(), 3);
+    assert_eq!(list_counter.count(), 3);
+}
+
+#[tokio::test]
+async fn test_lookup_then_open_cached() {
+    let fs_config = S3FilesystemConfig {
+        cache_config: CacheConfig {
+            prefer_s3: false,
```

Comment:
nit: this field was weirdly named — it's not really a preference? it's whether metadata can be served from cache or must be refreshed?

Reply:
I've gone through a lot of names with it that I'm not super happy with, and this is no exception. :)

OK, I think the way you worded it is quite nice. `getattr` unfortunately is a bit of a weird case.

Maybe `serve_from_cache`? All operations will be served from cache, except `getattr`. Not perfect but maybe close enough.

Reply:
`serve_lookup_from_cache` seems good enough, obviously it affects `open` too but maybe not worth trying to shove into the name

Reply:
Addressing in #573.